src/Symfony/Component/DependencyInjection/Compiler/AutowirePass.php

```diff
@@ -273,6 +273,12 @@ private function autowireMethod(\ReflectionFunctionAbstract $reflectionMethod, a
 
                 // no default value? Then fail
                 if (!$parameter->isDefaultValueAvailable()) {
+                    // For core classes, isDefaultValueAvailable() can
+                    // be false when isOptional() returns true. If the
+                    // argument *is* optional, allow it to be missing
+                    if ($parameter->isOptional()) {
+                        continue;
```

Comment:
There could be holes with `break` or `continue`. The already-specified `$arguments` are passed into this function (imagine that thanks to named arguments, someone has specified the 1st and 4th arguments to something (i.e. $arguments has`0` and `3` indexes). If we hit this code when trying to autowire the 2nd argument (index `1`), then a `break` would still mean that `$arguments` ultimately has indexes `0` and `3` (the same with a `continue`: continue would loop again, but not fill in any args).

So I think changing to `break` makes sense... but we still have holes. This is caught later by `CheckArgumentsValidityPass` with:

> Invalid constructor argument 4 for service "foo_bar": argument 2 must be defined before. Check your service definition.

tl;dr This DOES allow for holes in the parameters list in this pass... but that problem is caught by a later pass.